modin/experimental/xgboost/xgboost_ray.py

```diff
@@ -0,0 +1,339 @@
+
+    def get_actor_ip(self):
+        return self._ip
+
+    def _get_dmatrix(self, *X_y):
+        s = time.time()
+        X = X_y[: len(X_y) // 2]
+        y = X_y[len(X_y) // 2 :]
+
+        assert (
+            len(X) == len(y) and len(X) > 0
+        ), "X and y should have the equal length more than 0"
+
+        X = pandas.concat(X, axis=0)
+        y = pandas.concat(y, axis=0)
+        LOGGER.info(f"Concat time: {time.time() - s} s")
+
+        return xgb.DMatrix(X, y)
+
+    def set_train_data(self, *X_y, add_as_eval_method=None):
+        self._dtrain = self._get_dmatrix(*X_y)
+
+        if add_as_eval_method is not None:
+            self._evals.append((self._dtrain, add_as_eval_method))
+
+    def set_predict_data(
+        self,
+        *X,
+    ):
+        for x in X:
+            self._dpredict.append(xgb.DMatrix(x, None))
+
+    def add_eval_data(self, *X_y, eval_method):
+        self._evals.append((self._get_dmatrix(*X_y), eval_method))
+
+    def train(self, rabit_args, params, *args, **kwargs):
+        local_params = params.copy()
+        local_dtrain = self._dtrain
+        local_evals = self._evals
+
+        local_params["nthread"] = self._nthreads
+
+        evals_result = dict()
+
+        s = time.time()
+        with RabitContext(self._ip, rabit_args):
+            bst = xgb.train(
+                local_params,
+                local_dtrain,
+                *args,
+                evals=local_evals,
+                evals_result=evals_result,
+                **kwargs,
+            )
+            LOGGER.info(f"Local training time: {time.time() - s} s")
+            return {"booster": bst, "history": evals_result}
+
+    def predict(self, booster: xgb.Booster, *args, **kwargs):
+        local_dpredict = self._dpredict
+        booster.set_param({"nthread": self._nthreads})
+
+        s = time.time()
+        predictions = [booster.predict(X, *args, **kwargs) for X in local_dpredict]
+        LOGGER.info(f"Local prediction time: {time.time() - s} s")
+        return np.concatenate(predictions)
+
+
+def create_actors(num_cpus=1, nthread=cpu_count()):
+    num_nodes = len(ray.nodes())
+
+    # Create remote actors
+    actors = {
+        node_info.split("node:")[-1]: ModinXGBoostActor.options(
+            num_cpus=num_cpus, resources={node_info: 1.0}
+        ).remote(node_info.split("node:")[-1], nthread=nthread)
+        for node_info in ray.cluster_resources()
+        if "node" in node_info
+    }
+
+    assert num_nodes == len(
+        actors
+    ), f"Number of nodes {num_nodes} is not equal to number of actors {len(actors)}."
+
+    return actors
+
+
+def _split_data_across_actors(
+    actors: Dict, set_func, X_parts, y_parts=None, evenly_data_distribution=True
+):
+    if not evenly_data_distribution:
+        actors_ips = list(actors.keys())
+        partitions_ips = [ray.get(row_part[0]) for row_part in X_parts]
+        if len(np.unique(actors_ips)) != len(np.unique(partitions_ips)):
```

Comment:
This check feels strange... take `actor_ips = [1]` and `partition_ips = [2]` - they have same amount of unique items but are actually different.
I think you should compute `empty_nodes_ips` here and check for it being non-empty.

Reply:
IPs of partitions is subset of set of actors IPs because the same ray cluster is used. Therefore, in your example `actor_ips = [1,2]` and check will work correctly. But it was changed using list of empty nodes.